MEN-4550: Add acceptance test for DBus Update Control endpoint.

kacf opened the pull request:
`mender-mock-server.py` has to be substantially extended for the calls
to work correctly. We have to consume the body of requests for
subsequent calls to work, and we need to be able to handle chunked
bodies. We also make it possible to schedule a deployment with it.

In addition we add a logger-update-module which we can use to track
exactly which states the Update Module is going through.

Changelog: None

Signed-off-by: Kristian Amlie <kristian.amlie@northern.tech>

lluiscampos commented on tests/acceptance/mock_server.py:
```diff
@@ -0,0 +1,117 @@
+#!/usr/bin/python
+# Copyright 2020 Northern.tech AS
```
2021.

I guess we don't have these checks on meta-mender... :memo:

lluiscampos commented on tests/acceptance/mock_server.py:
```diff
@@ -0,0 +1,117 @@
+)
+
+
+@pytest.fixture
```
Better to be explicit, I think
```suggestion
@pytest.fixture(scope="function")
```

lluiscampos commented on tests/acceptance/mock_server.py:
```diff
@@ -0,0 +1,117 @@
+    result = connection.run(
+        f"mkdir -p {bdir} && if [ -e {conffile} ]; then cp {conffile} {conffile_bkp}; fi"
+    )
+    assert result.exited == 0
```
(Just because I have seen this often in modern tests)

In client acceptance tests, a `connection.run` command not exiting with `0` will raise an `Exception` and finish the test, so asserting here is not really adding value. If you want to run a command for which we can expect a non-0 return, then we add the `warn=True` parameter.

kacf replied:
Right. This is preexisting, but I can remove it.

lluiscampos commented on tests/meta-mender-ci/recipes-testing/logger-update-module/logger-update-module.bb:
```diff
@@ -0,0 +1,30 @@
+LIC_FILES_CHKSUM = "file://${COREBASE}/meta/files/common-licenses/Apache-2.0;md5=89aea4e17d99a7cacdbeed46a0096b10"
```
This is equivalent to:
```suggestion
LIC_FILES_CHKSUM = "file://${COMMON_LICENSE_DIR}/Apache-2.0;md5=89aea4e17d99a7cacdbeed46a0096b10"
```

Which we are already using in other recipes

lluiscampos commented on tests/meta-mender-ci/recipes-testing/logger-update-module/logger-update-module.bb:
```diff
@@ -0,0 +1,30 @@
+do_install() {
+    install -m 755 -d ${D}${datadir}/mender/modules/v3
+    install -m 755 ${B}/logger-update-module ${D}${datadir}/mender/modules/v3/logger-update-module
+}
```
Nice! :+1:

lluiscampos commented on tests/acceptance/test_update_control.py:
```diff
@@ -0,0 +1,431 @@
+    )
+    def test_update_control_maps(
+        self,
+        case_name,
```
`case_name` is unused

kacf replied:
It is required by `parametrize`, otherwise it complains about a missing parameter. The ultimate goal is that the test case be named instead of just an index, which is much harder to identify in the list of cases.

lluiscampos commented on tests/acceptance/test_update_control.py:
```diff
@@ -0,0 +1,431 @@
+                "last_successful_state": "ArtifactVerifyReboot",
+            },
+        },
+    ]
```
Excellent tests. I was able to follow all the cases with no much trouble.

kacf replied:
That's great! I was afraid the many labels would be a bit too much. At one point did I wonder if it should be split, but there is a lot of reuse.

lluiscampos reviewed (approved):
I really couldn't find find much, see couple of minor comments below.

The test code is very good, the server mock code a tiny bit cumbersome, but I didn't had better suggestions neither.

Well done! Both with the tests and with the bugs that we have revealed (and fixed!).

kacf commented:
> the server mock code a tiny bit cumbersome, but I didn't had better suggestions neither.

Agree on both counts!

Thanks for the review, I have fixed all the issues pointed out, as well as (hopefully) the test failures. All were minor things, so I squashed.

oleorhagen commented on tests/acceptance/test_update_control.py:
```diff
@@ -0,0 +1,432 @@
+            continue_map_inserted = False
+            second_deployment_done = False
+            PAUSE_STATE_OBSERVE_COUNT = 2
+            while time.time() - now <= case.get("fail_after", FAIL_TIME):
```
Clever

oleorhagen reviewed (approved):
Nice!

I wasn't really able to get into the matter too deep, but on the surface looks real nice

kacf commented on tests/acceptance/mock_server.py:
```diff
@@ -0,0 +1,116 @@
+    }
+}
+""" % (
+    EXPIRATION_TIME,
```
That's what I started with, but it doesn't work because it gets confused by the `{}` opening and closing braces for JSON.
